### PR TITLE
Merge aw-watcher-bash-preexec into aw-watcher-bash.sh

### LINKED ISSUE #5: Merge aw-watcher-bash.sh and aw-watcher-bash-preexec.sh
As mentioned [here](https://github.com/Otto-AA/aw-watcher-bash/pull/4#issue-199800086) it would be nice for the install process to merge aw-watcher-bash.sh and aw-watcher-bash-preexec.sh

The problem I face with this is, that the code of aw-watcher-bash.sh partly depends on being executed by bash. Therefore, if we merge it and then put it (for instance) into .zshrc, it will break when it gets executed by zsh. In particular arrays are treated differently in these two shells (bash has them zero-indexed, zsh one-indexed).

Do you have an idea, how we could fix this, @johan-bjareholt ?

If not, I would keep these files separated.

### PR DESCRIPTION
**Pull request by johan-bjareholt**
Works, but have not tested properly yet and still don't fully understand the implementation so there is likely multiple errors.
Don't have time to work on it anymore today (but hopefully will tomorrow) so I open this review to let @Otto-AA be able to see the changes in the mean time.

Fixes #5

### CODE REVIEW

**Review by Otto-AA — CHANGES_REQUESTED** [review]
Seems to work fine :)

Here are some changes I would make. If you don't agree with them, just say it. Some are maybe personal preferences.

Here is the file in the edited version: https://repl.it/repls/AthleticShallowFactorial

**Comment by Otto-AA** [comment]
Btw, I won't be able to code for a few weeks, starting next Tuesday. So don't be surprised if I don't react to pull-requests/...

File: aw-watcher-bash.sh [hunk 1]
```diff
@@ -1,27 +1,31 @@
...
-# Prepends a backslash to every quote (") and every backslash followed by a quote (\") or backslash (\\)
-#
-# From: echo "\"" > "output"
-# To:   echo \"\\\"\" > \"output\"
-# From: echo "\\" > "output"
-# To:   echo \"\\\\\" > \"output\"
-escape_quotes() {
+export PATH=$PATH:~/.local/bin/
+
+current_shell=$(ps -p "$$" | tail -n 1 | awk '{ print $4 }')
```
[the remark this answers is not in the document]
Comment (johan-bjareholt) [reply aw-watcher-bash.sh]: Fixed

File: aw-watcher-bash.sh [hunk 2]
```diff
@@ -1,27 +1,31 @@
...
+  # zsh supports preexec natively, do nothing
+  : # : is a nop
+elif [[ -f ~/.bash-preexec.sh ]]; then
+  source ~/.bash-preexec.sh
+else
+  echo "preexec is not supported by your terminal and bash-preexec is not installed at ~/.bash-preexec.sh, aw-watcher-bash will not work"
+fi
+
+
+function escape_quotes() {
```
[the remark this answers is not in the document]
Comment (johan-bjareholt) [reply aw-watcher-bash.sh]: Fixed

File: aw-watcher-bash.sh [hunk 3]
```diff
@@ -33,71 +37,85 @@ escape_quotes() {
...
-        add_arg '--command'
-        ;;
-    'precmd')
-        add_arg '--exit-code'
-        ;;
-    'preclose')
-        ;;
-    *)
-        echo "Unknown event: $event" >> "$logfile"
+function send_event() {
```
[the remark this answers is not in the document]
Comment (johan-bjareholt) [reply aw-watcher-bash.sh]: Fixed

File: aw-watcher-bash.sh [hunk 4]
```diff
@@ -33,71 +37,85 @@ escape_quotes() {
...
+    # for an accurate version of the parameters see add_arg
+    pipe_message_args=()
+    event=$1
+    pid=$2
+    extra_arg=$3
+
+    logfile="${XDG_CACHE_HOME:-$HOME/.cache}/activitywatch/log/aw-watcher-terminal/aw-watcher-bash.log"
+    fifo_path="${XDG_DATA_HOME:-$HOME/.local/share}/activitywatch/aw-watcher-terminal/aw-watcher-terminal-fifo"
+
+    function add_arg() {
```
[the remark this answers is not in the document]
Comment (johan-bjareholt) [reply aw-watcher-bash.sh]: Fixed

File: aw-watcher-bash.sh [hunk 5]
```diff
@@ -33,71 +37,85 @@ escape_quotes() {
...
+        # $1: key
+        # $2: val (or --first-val)
+        # $3: --shift or --no-shift  (use this to remove the latest value)
+        key="$1"
+        val="$2"
+        pipe_message_args+=("$key \"$(escape_quotes "$val")\"")
+    }
+
+    add_arg '--event' "$event"
+    add_arg '--pid' "$2"
```
[the remark this answers is not in the document]
Comment (johan-bjareholt) [reply aw-watcher-bash.sh]: Fixed

File: aw-watcher-bash.sh [hunk 6]
```diff
@@ -33,71 +37,85 @@ escape_quotes() {
...
+}
 
-esac
+send_aw_watcher_bash_event() {
+    # Call send_event in a background process to
+    # prevent blocking and in a subshell to prevent logging
+    (send_event "$1" "$PPID" "$2" &)
+}
 
+aw_watcher_terminal_preclose() {
```
[the remark this answers is not in the document]
Comment (johan-bjareholt) [reply aw-watcher-bash.sh]: Fixed

File: aw-watcher-bash.sh [hunk 7]
```diff
@@ -33,71 +37,85 @@ escape_quotes() {
...
+    send_aw_watcher_bash_event 'preclose'
+}
 
-message="${pipe_message_args[*]} --send-heartbeat"
+send_aw_watcher_bash_event 'preopen'
+trap aw_watcher_terminal_preclose TERM EXIT
 
-# send message to pipe and logfile
-echo "$message" > "$fifo_path" &
-echo "$message" >> "$logfile"
```
[the remark this answers is not in the document]
Comment (johan-bjareholt) [reply aw-watcher-bash.sh]: Just thought it was unnecessary, might become large after a few months. We could possibly keep it commented out though to easily enable debugging.

**Comment by johan-bjareholt** [comment]
> Btw, I won't be able to code for a few weeks, starting next Tuesday. So don't be surprised if I don't react to pull-requests/...

Alright, thanks for all the great work!
I hope we will be able to work some more later on :)

File: aw-watcher-bash.sh (hunk 7 above)
[the remark this answers is not in the document]
Comment (Otto-AA) [reply aw-watcher-bash.sh]: Good idea :)

**Comment by Otto-AA** [comment]
> Alright, thanks for all the great work!
I hope we will be able to work some more later on :)

And thank you for the feedback you gave me 👍 
As this is my first project I'm working on together with others, it is really appreciated.

**Review by Otto-AA — APPROVED** [review]

File: aw-watcher-bash.sh [hunk 8]
```diff
@@ -1,103 +1,94 @@
...
+        local _escaped="${1/\\/\\\\}"
+        _escaped="${_escaped/\"/\\\"}"
+        echo "$_escaped"
+    }
+
+    function add_arg() {
+    # Add an argument (e.g. --pid 1234) to the pipe_message_args array
+    # $1: key
+    # $2: val (or --first-val)
+    pipe_message_args+=("$1 \"$(escape_quotes "$2")\"")
```
Comment (Otto-AA) [inline aw-watcher-bash.sh]: Would be nice if you could intend it a little bit more :)

**Comment by Otto-AA** [comment]
Is there still something you want to change (as it is "WIP") or is it ready to be merged?

**Comment by johan-bjareholt** [comment]
Don't think so no

**Pull request merged**